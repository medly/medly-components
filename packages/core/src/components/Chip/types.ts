import { HTMLProps, Omit } from '@medly-components/utils';

type ButtonProps = Omit<HTMLProps<HTMLButtonElement>, 'type'>;

export interface StyledProps extends ButtonProps {
    /** Chip design */
    variant?: 'solid' | 'flat' | 'outlined';
    /** Button colors */
    color?: string;
    /** To be used to render any component as Chip */
    as?: any;
    /** To be used if rendering like Link */
    to?: string;
}

export interface ChipProps extends StyledProps {
    /** Chip label */
    label: string;
    /** On click function */
    onClick?: () => void;
    /** On delete function */
    onDelete?: () => void;
}
