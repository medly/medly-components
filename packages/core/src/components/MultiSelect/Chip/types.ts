import { HTMLProps, Omit, WithThemeProp } from '@medly-components/utils';

type ButtonProps = Omit<HTMLProps<HTMLButtonElement>, 'type' | 'label'>;

export interface Props extends Omit<ButtonProps, 'size'>, WithThemeProp {
    size: 'S' | 'M';
    label: number | string;
    variant?: 'outlined' | 'filled' | 'fusion';
    state?: 'default' | 'active' | 'error' | 'disabled';
    onClear?: (label?: number | string) => void;
}
