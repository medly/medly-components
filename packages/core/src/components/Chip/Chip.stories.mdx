import { Chip } from './Chip';
import { updateNestedValue } from '@medly-components/utils';
import { ThemeProvider } from 'styled-components';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, color, select, boolean, text } from '@storybook/addon-knobs';
import * as stories from './Chip.stories';
import { defaultTheme } from '@medly-components/theme';
import { action } from '@storybook/addon-actions';

<Meta title="Core" component={Chip} />

# Chip

Chip represents complex entities in small blocks, such as a contact or a location.

Use it to:
-   Tag by property
-   Categorize the selected items from the dropdown list
-   Highlight an item status for quick recognition


### Basic Chip

<Preview withToolbar>
    <Story name="Chip" decorators={[withKnobs]}>
        <Chip
            label={text('Label', 'Demo')}
            variant={select('Variant', stories.variant, 'solid')}
            onClick={action('Chip Clicked')}
            onDelete={action('Delete Clicked')}
            disabled={boolean('Disabled', false)}
            color={color('Color', defaultTheme.chip.defaultColor)}
        />
    </Story>
</Preview>

### Variants

Changes the style of the button.

<Preview>
    <Chip label="Solid Chip" onDelete={action('Delete Clicked')} variant="solid" color={defaultTheme.chip.defaultColor} />
    <Chip label="Outlined Chip" onDelete={action('Delete Clicked')} variant="outlined" color={defaultTheme.chip.defaultColor} />
    <Chip label="Flat Chip" onDelete={action('Delete Clicked')} variant="flat" color={defaultTheme.chip.defaultColor} />
</Preview>

### Disabled

Disables the style for all the variants.

<Preview>
    <Chip disabled label="Solid Chip" variant="solid" color={defaultTheme.chip.defaultColor} />
    <Chip disabled label="Outlined Chip" variant="outlined" color={defaultTheme.chip.defaultColor} />
    <Chip disabled label="Flat Chip" variant="flat" color={defaultTheme.chip.defaultColor} />
</Preview>

### Delete Icon

Use the `onDelete` prop to feature a delete 'x' icon on the chip.

<Preview withToolbar>
    <Chip label="Chip without delete icon" onClick={action('Clicked')} color={defaultTheme.chip.defaultColor} />
</Preview>

### Custom Colors

To use custom colors, provide the `custom` theme to the `chip` component. Use the `updateNestedValue` function available in `@medly-components/utils` to update the nested values in the `defaultTheme` object. Check out the `ChipTheme` section to find out the customizable properties.

```tsx
import { updateNestedValue } from '@medly-components/utils';
import { defaultTheme } from '@medly-components/theme';
import { ThemeProvider } from 'styled-components';
import { Chip } from './Chip.tsx';

<ThemeProvider
    theme={updateNestedValue(defaultTheme, 'chip', {
        defaultColor: defaultTheme.colors.red[500],
        disabledColor: defaultTheme.colors.grey[200]
    })}
>
    <Chip label="Red Chip" variant="solid" />
</ThemeProvider>;
```

<Preview>
    <ThemeProvider
        theme={updateNestedValue(defaultTheme, 'chip', {
            defaultColor: defaultTheme.colors.red[500],
            disabledColor: defaultTheme.colors.grey[200]
        })}
    >
        <Chip label="Red Chip" />
    </ThemeProvider>
</Preview>

### Props

<Props of={Chip} />

### ChipTheme

<Props of={stories.ThemeInterface} />
