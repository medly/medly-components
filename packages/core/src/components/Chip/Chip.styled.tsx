import { SvgIcon } from '@medly-components/icons';
import { centerAligned, css, styled, WithThemeProp } from '@medly-components/utils';
import Text from '../Text';
import { StyledProps } from './types';

const getColor = ({ theme, color, disabled }: StyledProps & WithThemeProp) =>
    disabled ? theme.chip.disabledColor : color || theme.chip.defaultColor;

const outlined = () => css`
    border: 1px solid ${getColor};
    color: ${getColor};
    ${SvgIcon} {
        * {
            fill: ${getColor};
        }
    }
`;

const solid = () => css`
    border: 1px solid ${getColor};
    background-color: ${getColor};
    color: ${({ theme }) => theme.colors.white};
    ${SvgIcon} {
        * {
            fill: ${({ theme }) => theme.colors.white};
        }
    }
`;

const flat = () => css`
    color: ${getColor};
    ${SvgIcon} {
        * {
            fill: ${getColor};
        }
    }
`;

export const Chip = styled('button')<StyledProps>`
    background-color: transparent;
    font-family: inherit;
    border: none;
    border-radius: 20px;
    padding: 3px 10px;
    margin: 2px;
    user-select: none;
    text-decoration: none;
    max-width: max-content;
    cursor: ${({ onClick, disabled }) => (disabled ? 'not-allowed' : onClick ? 'pointer' : 'default')};

    &:focus {
        outline: none;
    }

    ${props => props.variant === 'flat' && flat()};
    ${props => props.variant === 'outlined' && outlined()};
    ${props => props.variant === 'solid' && solid()};

    ${Text.Style} + ${SvgIcon}{
        cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
        margin-left: 10px;
    }

    ${centerAligned()}
`;
