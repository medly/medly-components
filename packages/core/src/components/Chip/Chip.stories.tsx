import { ChipTheme, defaultTheme } from '@medly-components/theme';
import { action } from '@storybook/addon-actions';
import { boolean, color, select, text } from '@storybook/addon-knobs';
import React from 'react';
import { Chip } from './Chip';
import { ChipProps } from './types';

export const variant: Required<ChipProps>['variant'][] = ['solid', 'flat', 'outlined'];

export const ThemeInterface: React.FC<ChipTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.chip
};

export const Basic = () => (
    <Chip
        label={text('Label', 'Demo')}
        variant={select('Variant', variant, 'solid')}
        onClick={action('Chip Clicked')}
        onDelete={action('Delete Clicked')}
        disabled={boolean('Disabled', false)}
        color={color('Color', defaultTheme.chip.defaultColor)}
    />
);
