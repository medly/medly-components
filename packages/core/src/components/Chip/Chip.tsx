import { ClearIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import React, { FC, useCallback } from 'react';
import Text from '../Text';
import * as Styled from './Chip.styled';
import { ChipProps } from './types';

const Component: FC<ChipProps> = React.memo(
    React.forwardRef((props, ref) => {
        const { label, onDelete, ...restProps } = props;

        const handleDelete = useCallback(
            (e: React.MouseEvent<HTMLOrSVGElement>) => {
                e.stopPropagation();
                !restProps.disabled && onDelete && onDelete();
            },
            [restProps.disabled, onDelete]
        );

        return (
            <Styled.Chip ref={ref} {...restProps}>
                <Text>{label}</Text>
                {onDelete && <ClearIcon size="S" onClick={handleDelete} title={`${label} chip clear icon`} />}
            </Styled.Chip>
        );
    })
);

Component.defaultProps = {
    variant: 'solid'
};
Component.displayName = 'Chip';
export const Chip: FC<ChipProps> & WithStyle = Object.assign(Component, { Style: Styled.Chip });
