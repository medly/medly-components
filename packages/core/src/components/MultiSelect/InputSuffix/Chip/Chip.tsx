import { ClearIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import React, { FC, useCallback } from 'react';
import Text from '../../../Text';
import * as Styled from './Chip.styled';
import { Props } from './types';

const Component: FC<Props> = React.memo(({ id, label, onClear, state, ...restProps }) => {
    const onClickHandler = useCallback((e: React.MouseEvent) => e.stopPropagation(), []),
        onClearHandler = useCallback(
            (e: React.MouseEvent<HTMLOrSVGElement>) => {
                e.stopPropagation();
                !restProps.disabled && onClear && onClear(label);
            },
            [restProps.disabled, onClear]
        );

    return (
        <Styled.Chip id={id} state={state!} {...restProps} onClick={onClickHandler}>
            <Text textWeight="Medium" textVariant={restProps.size === 'S' ? 'body2' : 'body1'}>
                {label}
            </Text>
            {!restProps.hideClearIcon && <ClearIcon id={`${id}-clear`} title={`${id}-clear-icon`} onClick={onClearHandler} />}
        </Styled.Chip>
    );
});

Component.displayName = 'Chip';
Component.defaultProps = { state: 'default', variant: 'outlined' };
export const Chip: FC<Props> & WithStyle = Object.assign(Component, { Style: Styled.Chip });
