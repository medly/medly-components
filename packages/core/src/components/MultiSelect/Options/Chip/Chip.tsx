import { ClearIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import type { FC } from 'react';
import { memo, useCallback } from 'react';
import Text from '../../../Text';
import * as Styled from './Chip.styled';
import { Props } from './types';

const Component: FC<Props> = memo(({ id, value, label, disabled, onClear }) => {
    const chipId = id ? id.replace(' ', '') : 'medly-options-chip',
        onClearHandler = useCallback(() => !disabled && onClear && onClear(value), [disabled, onClear, value]);

    return (
        <Styled.Chip id={chipId} disabled={disabled}>
            <Text textVariant="button2">{label}</Text>
            <ClearIcon id={`${chipId}-clear`} title={`${chipId}-clear-icon`} onClick={onClearHandler} />
        </Styled.Chip>
    );
});

Component.displayName = 'Chip';
export const Chip: FC<Props> & WithStyle = Object.assign(Component, { Style: Styled.Chip });
