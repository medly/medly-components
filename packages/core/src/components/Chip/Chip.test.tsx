import { cleanup, fireEvent, render, screen } from '@test-utils';
import React from 'react';
import { Chip } from './Chip';

describe('Chip component', () => {
    afterEach(cleanup);

    it('should render properly with flat variant', () => {
        const { container } = render(<Chip variant="flat" color="yellow" label="Flat Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly with outlined variant', () => {
        const { container } = render(<Chip variant="outlined" color="secondary" label="Outlined Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly with solid variant', () => {
        const { container } = render(<Chip variant="solid" label="Solid Chip" />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly when it is disabled', () => {
        const { container } = render(<Chip disabled variant="solid" label="Solid Chip" />);
        expect(container).toMatchSnapshot();
    });

    test.each([true, false])('should handle on click when chip is disabled: %p', (disabled: boolean) => {
        const mockOnClick = jest.fn();

        const { getByText } = render(<Chip disabled={disabled} variant="flat" color="yellow" onClick={mockOnClick} label="Flat Chip" />);
        fireEvent.click(getByText('Flat Chip'));
        expect(mockOnClick).toHaveBeenCalledTimes(disabled ? 0 : 1);
    });

    test.each([true, false])('should handle on delete when chip is disabled: %p', (disabled: boolean) => {
        const mockOnClick = jest.fn(),
            mockOnDelete = jest.fn();

        render(<Chip disabled={disabled} variant="flat" onClick={mockOnClick} onDelete={mockOnDelete} color="yellow" label="Dummy" />);
        fireEvent.click(screen.getByTitle('Dummy chip clear icon'));
        expect(mockOnDelete).toHaveBeenCalledTimes(disabled ? 0 : 1);
        expect(mockOnClick).not.toBeCalled();
    });
});
