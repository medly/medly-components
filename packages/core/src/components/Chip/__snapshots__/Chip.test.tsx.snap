// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip component should render properly when it is disabled 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  background-color: transparent;
  font-family: inherit;
  border: none;
  border-radius: 20px;
  padding: 3px 10px;
  margin: 2px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  cursor: not-allowed;
  border: 1px solid #b0bcc8;
  background-color: #b0bcc8;
  color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:focus {
  outline: none;
}

.c0 .sc-1pf3x8a-0 * {
  fill: #ffffff;
}

.c0 .c1 + .sc-1pf3x8a-0 {
  cursor: not-allowed;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    <span
      class="c1 c2"
    >
      Solid Chip
    </span>
  </button>
</div>
`;

exports[`Chip component should render properly with flat variant 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  background-color: transparent;
  font-family: inherit;
  border: none;
  border-radius: 20px;
  padding: 3px 10px;
  margin: 2px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  cursor: default;
  color: yellow;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:focus {
  outline: none;
}

.c0 .sc-1pf3x8a-0 * {
  fill: yellow;
}

.c0 .c1 + .sc-1pf3x8a-0 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
    color="yellow"
  >
    <span
      class="c1 c2"
    >
      Flat Chip
    </span>
  </button>
</div>
`;

exports[`Chip component should render properly with outlined variant 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  background-color: transparent;
  font-family: inherit;
  border: none;
  border-radius: 20px;
  padding: 3px 10px;
  margin: 2px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  cursor: default;
  border: 1px solid secondary;
  color: secondary;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:focus {
  outline: none;
}

.c0 .sc-1pf3x8a-0 * {
  fill: secondary;
}

.c0 .c1 + .sc-1pf3x8a-0 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
    color="secondary"
  >
    <span
      class="c1 c2"
    >
      Outlined Chip
    </span>
  </button>
</div>
`;

exports[`Chip component should render properly with solid variant 1`] = `
.c2 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
}

.c0 {
  background-color: transparent;
  font-family: inherit;
  border: none;
  border-radius: 20px;
  padding: 3px 10px;
  margin: 2px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  cursor: default;
  border: 1px solid #126AFA;
  background-color: #126AFA;
  color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:focus {
  outline: none;
}

.c0 .sc-1pf3x8a-0 * {
  fill: #ffffff;
}

.c0 .c1 + .sc-1pf3x8a-0 {
  cursor: pointer;
  margin-left: 10px;
}

<div>
  <button
    class="c0"
  >
    <span
      class="c1 c2"
    >
      Solid Chip
    </span>
  </button>
</div>
`;
